Python/errors.c

```diff
@@ -1287,12 +1287,13 @@ write_unraisable_exc_file(PyThreadState *tstate, PyObject *exc_type,
     }
 
     assert(PyExceptionClass_Check(exc_type));
-    const char *className = PyExceptionClass_Name(exc_type);
-    if (className != NULL) {
-        const char *dot = strrchr(className, '.');
-        if (dot != NULL) {
-            className = dot+1;
-        }
+    PyObject *qualName = PyType_GetQualName((PyTypeObject *)exc_type);
```

Comment:
Looks like there's a refleak here; `PyType_GetQualName` returns a new ref.

Reply:
> that better than me :) IIRC, the old code

I see it clears in the end of the function if err < 0. So as long as we don't reset err it should be ok.

Reply:
Actually it's mixed - sometimes it clears and sometimes sets err to -1 and sometimes both.   

I don't know why the Clear at the end is under if (err < 0) rather than if (PyErr_Occurred()).

Reply:
Ok, I think I get it. If we have an error trying to figure out the classname then we print "<unknown>" so we've recovered from the error so we need to clear it and not set err to -1. 

err < 0 in the end should be because some PyFile_WriteString has failed.

Reply:
Also - once I move the code to where it's used it becomes clear that we don't need the classname variable at all, we can just print qualname.

Reply:
> Ok, I think I get it. If we have an error trying to figure out the classname then we print "<unknown>" so we've recovered from the error so we need to clear it and not set err to -1.

Yes, that makes sense.

> Also - once I move the code to where it's used it becomes clear that we don't need the classname variable at all, we can just print qualname.

Neat!